presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java

```diff
@@ -0,0 +1,231 @@
+        }
+
+        private RowExpression randomizeJoinKey(RowExpression keyExpression, String prefix)
+        {
+            int partitionCount = getHashPartitionCount(session);
+            RowExpression randomNumber = call(
+                    functionAndTypeManager,
+                    "random",
+                    BIGINT,
+                    constant((long) partitionCount, BIGINT));
+            RowExpression randomNumberVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, randomNumber.getType(), VARCHAR), VARCHAR, randomNumber);
+            RowExpression concatExpression = call(functionAndTypeManager,
+                    "concat",
+                    VARCHAR,
+                    ImmutableList.of(constant(Slices.utf8Slice(prefix), VARCHAR), randomNumberVarchar));
+
+            RowExpression castToVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, keyExpression.getType(), VARCHAR), VARCHAR, keyExpression);
+            return new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, VARCHAR, ImmutableList.of(castToVarchar, concatExpression));
+        }
+
+        // Do not need to generate randomized variable if the joinKey 1) has already been randomized 2) included in the output of the source node
+        private boolean isAlreadyRandomized(PlanNode source, VariableReferenceExpression joinKey)
+        {
+            return keyToRandomKeyMap.containsKey(joinKey) && source.getOutputVariables().contains(keyToRandomKeyMap.get(joinKey));
+        }
+
+        private Map<VariableReferenceExpression, RowExpression> generateRandomKeyMap(List<VariableReferenceExpression> joinKeys, String prefix)
+        {
+            List<RowExpression> randomExpressions = joinKeys.stream().map(x -> randomizeJoinKey(x, prefix)).collect(Collectors.toList());
+            List<VariableReferenceExpression> randomVariable = randomExpressions.stream().map(x -> planVariableAllocator.newVariable(x, RandomizeNullKeyInOuterJoin.class.getSimpleName())).collect(Collectors.toList());
+
+            checkState(joinKeys.size() == randomVariable.size());
+            Map<VariableReferenceExpression, VariableReferenceExpression> newKeyToRandomMap = IntStream.range(0, joinKeys.size()).boxed()
+                    .collect(ImmutableMap.toImmutableMap(i -> joinKeys.get(i), i -> randomVariable.get(i)));
+            keyToRandomKeyMap.putAll(newKeyToRandomMap);
+
+            Map<VariableReferenceExpression, RowExpression> result = IntStream.range(0, randomVariable.size()).boxed()
+                    .collect(ImmutableMap.toImmutableMap(i -> randomVariable.get(i), i -> randomExpressions.get(i)));
+
+            return result;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Void> context)
+        {
+            // For cross join, it does not allow mismatch of input and output. Need to reassign output in case input of the cross join changes.
+            if (joinNode.isCrossJoin()) {
```

Comment:
Can you explain more on why is this needed(what causes the mismatch)? Should we add a project on top of our rewrite, so we don't have to handle this?

Reply:
> Can you explain more on why is this needed(what causes the mismatch)?

This is because we also add the random key to the output of the optimized join node, so that it can be reused if in later stage we do outer join on the same join key.

The cross join's input can be changed if one of its input contains the outer join which are optimized by this rule.

> This seems like it could break in other places too.

Should not. In this rewrite, we are adding to the output, and it will be pruned if not used by later stages.

> Should we add a project on top of our rewrite, so we don't have to handle this?

If we add a project on top the rewrite to project off the random key added, we will not be able to reuse the random key at all.

Another way to do it is to add project node between the cross join and its source nodes, and project off the random key. This way the random key will not be available after the cross join and not able to reuse after it.

Reply:
Thanks for the detailed explanation, lets add a test for this.

Reply:
Tests added for cross join case.